crates/tbdex/src/messages/rfq.rs

```diff
@@ -12,48 +18,117 @@ pub struct Rfq {
+        let rfq = serde_json::from_str::<Self>(json)?;
+        rfq.verify()?;
+        Ok(rfq)
+    }
+
+    pub fn verify(&self) -> Result<()> {
+        Ok(crate::signature::verify(
+            &self.metadata.from,
+            &serde_json::to_value(self.metadata.clone())?,
+            &serde_json::to_value(self.data.clone())?,
+            &self.signature,
+        )?)
+    }
+
+    pub fn to_json(&self) -> Result<String> {
+        Ok(serde_json::to_string(&self)?)
     }
 
-    pub fn verify_offering_requirements(&self, _offering: Offering) -> Result<bool> {
+    pub fn verify_offering_requirements(&self, offering: &Offering) -> Result<bool> {
+        if offering.metadata.protocol != self.metadata.protocol {
+            return Err(MessageError::OfferingVerification(format!(
+                "offering has protocol version {} but rfq has protocol version {}",
+                offering.metadata.protocol, self.metadata.protocol
+            )));
+        }
+
+        if offering.metadata.id != self.data.offering_id {
+            return Err(MessageError::OfferingVerification(format!(
+                "offering id is {} but rfq has offering id {}",
+                offering.metadata.id, self.data.offering_id
+            )));
+        }
+
+        let payin_amount = self.data.payin.amount.parse::<f64>().map_err(|_| {
+            MessageError::OfferingVerification(format!(
+                "rfq payin amount invalid decimal string {}",
+                self.data.payin.amount
+            ))
+        })?;
+
+        if let Some(max_amount) = offering.data.payin.max.as_ref() {
+            let max_amount = max_amount.parse::<f64>().map_err(|_| {
+                MessageError::OfferingVerification(format!(
+                    "offering max amount invalid decimal string {}",
+                    max_amount
+                ))
+            })?;
+
+            if payin_amount > max_amount {
+                return Err(MessageError::OfferingVerification(format!(
+                    "rfq payin of {} is larger than max offering amount of {}",
+                    payin_amount, max_amount
+                )));
+            }
+        }
+
+        if let Some(min_amount) = offering.data.payin.min.as_ref() {
+            let min_amount = min_amount.parse::<f64>().map_err(|_| {
+                MessageError::OfferingVerification(format!(
+                    "offering min amount invalid decimal string {}",
+                    min_amount
+                ))
+            })?;
+
+            if payin_amount < min_amount {
+                return Err(MessageError::OfferingVerification(format!(
+                    "rfq payin of {} is smaller than min offering amount of {}",
+                    payin_amount, min_amount
+                )));
+            }
+        }
+
+        // verify payin
+        // verify payout
+        // verify claims
+
         println!("Rfq.verify_offering_requirements() invoked");
```

Comment:
```suggestion

```

Reply:
yeah, I'm working on this next, happy to remove for now